### PR TITLE
Add the posibility to specify more complex autosign_grains

### PR DESCRIPTION
**Pull request by johnvox**
### What does this PR do?
This PR allow admin to specify more complex autosign_grains.
The default behavior is preserved but if the grains is not found and the grain / autosign_grain match a pattern it will look into subkey

### What issues does this PR fix or reference?
#48776 - Nested grain in autosign_grains 
### Previous Behavior
Look only in first level grain key

### New Behavior
Loop over the grain data structure

### Tests written?
Yes

### Commits signed with GPG?
Yes

### Documentation written ?
Yes

### Local Test results
```
 python tests/runtests.py -n integration.daemons.test_masterapi.AutosignGrainsTest
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 3.6.8 (default, Aug 7 2019, 17:28:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Transplanting multimaster configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /opt/github/salt
 * Test suite is running under PID 19370
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Max open files settings is too low (soft: 1024, hard: 4096) for running the tests
 * Trying to raise the limits to soft: 3072, hard: 4096
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Setting up Salt daemons to execute tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Starting salt-master ... STARTED!
 * Starting salt-minion ... STARTED!
 * Starting sub salt-minion ... STARTED!
 * Starting syndic salt-master ... STARTED!
 * Starting salt-syndic ... STARTED!
===============================================================================================================================================================================================================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting integration.daemons.test_masterapi.AutosignGrainsTest Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.........................
----------------------------------------------------------------------
Ran 25 tests in 357.587s

OK

===========================================================================================  Overall Tests Report  ============================================================================================
***  No Problems Found While Running Tests  *******************************************************************************************************************************************************************
===============================================================================================================================================================================================================
OK (total=25, skipped=0, passed=25, failures=0, errors=0)
===========================================================================================  Overall Tests Report  ============================================================================================
Test suite execution finalized with exit code: 0
```

### CODE REVIEW

**Review by Akm0d — DISMISSED** [review]

File: doc/man/salt.7 [hunk 1]
```diff
@@ -39706,7 +39706,10 @@ New in version 2018.3.0.
 .sp
 To automatically accept minions based on certain characteristics, e.g. the \fBuuid\fP
 you can specify certain grain values on the salt master. Minions with matching grains
-will have their keys automatically accepted.
+will have their keys automatically accepted. The grain can contain a ':' character in
```
Comment (Ch3LL) [inline doc/man/salt.7]: we should not be editing the `man` pages directly as these will be generated from the docs each release. This should generate from the doc/topics/tutorials/autoaccept_grains.rst file.

File: salt/crypt.py [hunk 2]
```diff
@@ -822,7 +822,40 @@ def minion_sign_in_payload(self):
...
+                # If the autosign_grain is directly accessible, use it.
+                # it avoids issues if the grain key itself contains columns
+                element = None
+                if grain in self.opts['grains']:
+                    log.trace('Autosign Grains - grain %s - Found in grain root', grain)
+                    element = self.opts['grains'].get(grain, None)
+                # Else try to split the the grain key with column
+                # In every case, if the autosign_grain could not be found anywhere, return None
+                else:
+                    current_level = self.opts['grains']
```
Comment (Ch3LL) [inline salt/crypt.py]: wondering if this code could be condensed down to using `salt.utils.data.traverse_dict_and_list`

For example, given a grains structure that looks like this:

```
 ch3ll@thecakeisalie  ~/git/salt   autosign_grains ●  cat /etc/salt/grains 
nest:
  one:
    two:
      - this
```

```
In [6]: salt.utils.data.traverse_dict_and_list(__grains__, 'nest:one:two', delimiter=':')                                                            
Out[6]: ['this']
```

```
In [6]: salt.utils.data.traverse_dict_and_list(__grains__, 'nest:one:two', delimiter=':')                                                            
Out[6]: ['this']
```

You can look at how we are currently doing this in the grains execution module in the get method as well.

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: doc/man/salt.7 (hunk 1 above)
Comment (jynolen) [reply doc/man/salt.7]: Removal of changes done

**Comment by jynolen** [comment]
Make all changes suggest by @Ch3LL

**Review by Ch3LL — APPROVED** [review]

**Comment by sbernhard** [comment]
would be great if someone can have a look at this great addition. I'm currently having the same issue that a nested value is not possible to be used as autosign_grain (which is pretty bad).

**Comment by Ch3LL** [comment]
re-run full all

**Review by Ch3LL — CHANGES_REQUESTED** [review]
I set the tests to re-run, because all of hte tests where previously failing. I don't seem to have permissions to make changes, so can you add a changelog entry like so https://docs.saltproject.io/en/latest/topics/development/changelog.html

**Comment by jynolen** [comment]
@Ch3LL Sorry issue on my side. I deletesource branch (sorry about that)
Can you please consider following PR instead #61376 (I checkout the PR with gh CLI and make it in individual branch) thanks

**Comment by Ch3LL** [comment]
Yes, I'll close this one in favor of https://github.com/saltstack/salt/pull/61376

**Pull request closed without merging**